### ISSUE TITLE
Collapsible does not always handle growing children

### ISSUE THREAD

**Issue by Bringer128** [Shopify/polaris-react#7318]
## Issue summary

Collapsible occasionally does not recompute the size of its children even if they change their size appropriately within the React event cycle.

## Expected behavior

As the children grow, all _expanded_ Collapsible containers show their children

## Actual behavior

<details>
<summary>Gif recording of last Collapsible element not resizing</summary>
<img src="https://user-images.githubusercontent.com/1164344/193340151-6ac0e3fc-cf6c-4d2b-a3ef-d2b96ac82ae4.gif"/>
</details>

## Steps to reproduce the problem

A reproduction scenario is [here](https://codesandbox.io/s/quiet-leaf-bsrk37?file=/App.js)

If you want to recreate the repro:
1. Build a component that grows every second
2. In another component, have a button to add a Collapsible with the growing element inside.
3. Then click this button any number of times; the last added Collapsible will not grow but the others will

## Reduced test case

See the problem here: https://codesandbox.io/s/quiet-leaf-bsrk37?file=/App.js

## Specifications

- Are you using the React components? (Y/N): Yes
- Polaris version number: 10.4.1
- Browser: Google Chrome
- Device: MacBook Pro (14-inch, 2021); Apple M1 Pro
- Operating System: MacOS Monterey 12.6 (21G115)

**Comment by Bringer128** [comment]
As I investigate this, it turns out that the component relies on `onTransitionEnd` firing to consider the "animation complete" and let the size of the component set to `max-height: none`. However if it is rendered as "open" to start with, then the event does not fire.

**Comment by Bringer128** [comment]
From investigation it looks like the problem is this change: https://github.com/Shopify/polaris/pull/6100

Specifically, this change incorrectly fixed a bug in Secondary Navigation Items where the Collapsible would not change size if the children changed. However the root cause of that bug is the use of `duration: 0ms` [as seen here](https://github.com/Shopify/polaris/blob/b0445cf9b036752062af215da57347ce4b8f6f17/polaris-react/src/components/Navigation/components/Item/components/Secondary/Secondary.tsx#L19) to prevent the transition from occurring. This would cause the `Collapsible` to get stuck in the `animating` state, as the browser would not do the transition, and the `onTransitionEnd` callback would not fire.

When in the `animating` state, the `max-height` is set to a fixed value and overflow is hidden. This hides the contents of an extended list, which caused the original bug.

Unfortunately, this caused a couple of bugs. One is that the re-measure process when the component is already open is a bit janky, which is why [this PR](https://github.com/Shopify/polaris/issues/6283) was made (see the janky animation [here](https://github.com/Shopify/core-issues/issues/40141).

The other is that the PR made a `measure -> animate` cycle happen on first render. However it neglects [this useEffect](https://github.com/Shopify/polaris/blob/b0445cf9b036752062af215da57347ce4b8f6f17/polaris-react/src/components/Collapsible/Collapsible.tsx#L86-L91) call which sets the `height` state variable. Entering the `measure -> animate` cycle in this state means that the `height` _does not change_ on the `div` which means the transition does not occur.

Why did the original PR author not notice this? Because in Polaris by default, a `<ul>` has `14px` margin on the top and bottom (which is how the Collapsible is used in `Navigation`). And due to a quirk in how the component measures `scrollHeight` the height changes _twice_ during the `measure -> animate` cycle. During the `measure` state and the "first render" [useEffect]((https://github.com/Shopify/polaris/blob/b0445cf9b036752062af215da57347ce4b8f6f17/polaris-react/src/components/Collapsible/Collapsible.tsx#L86-L91)) the `max-height` is first measured when `overflow:visible` is set. Then in the next render `overflow:hidden` is set, which then includes the extra margin in the `scrollHeight`. Because the `height` changes between the first render and the `animating` state, a transition does occur and `onTransitionEnd` does fire. By contrast, the example in this issue _does not_ have a top and bottom margin, so gets stuck in the animating state.

TL;DR: The linked PR introduced this bug by trying to re-measure on children change. The correct fix is instead to avoid animating if `duration: 0ms` is used.

**Comment by Bringer128** [comment]
Note: If we choose to add a feature for "animate while open" it will need more thought. The way this component is designed today only handles animating between "open" and "closed" states.
